engine/grid.py

```diff
@@ -303,70 +304,106 @@ def find_border_nodes(self):
         cols = self.nodes.shape[1]
         rows = self.nodes.shape[0]
         leftmost_node = None
+        rightmost_node = None
         leftmost_node_pos = None
+        rightmost_node_pos = None
+        
         for row in range(rows):
             for col in range(cols):
                 node = self.nodes[row][col]
+                #border node 1 left active node 
                 if node.is_active and self.is_on_border(row, col, rows-1, cols-1):
                     # check if this is an active node and on the border
                     self.nodes[row][col].is_border = True
                     border_list.append((node, row, col))
                     if leftmost_node_pos is None or col < leftmost_node_pos[1]:
                         leftmost_node = node
                         leftmost_node_pos = (row, col)
+                if node.is_active and self.is_on_border(row, col, rows , cols):
+                    # check if this is an active node and on the border
+                    self.nodes[row][col].is_border = True
+                    if rightmost_node_pos is None or row > rightmost_node_pos[0]:
```

Comment:
if the current node has a row attribute i think changing this to 
if rightmost_node_pos is None or node.row > rightmost_node_pos[0] would be a little clearer, but not the biggest deal

Reply:
Actually the row is not an attribute, we are just iterating from 0 to the total number of rows. If this is confusing we can name it something else?

Reply:
Actually row is not an attribute, we are iterating from 0 to rows, and row is the counter :D